_posts/2016-06-30-state-in-scalable-architectures.md

```diff
@@ -0,0 +1,173 @@
+
+>> Rule 1: Each actor processes the messages from its inbox sequentially (not concurrently). So a clerk never picks two or more messages and works them side-by-side.
+
+>> Rule 2: If an actor sends to another actor two messages in a particular order then the receiving actor will receive them in the same order. So clerks can outsource work to their colleagues but again, they must take one message after the other out of the inbox and transfer them one after the other to the colleague.
+
+>> Rule 3: All messages sent to actors must be immutable objects.
+
+Every actor keeps its own state, but the processing happens sequentially, so no locking is required. This, plus the message delivery order guarantees and the immutability of the exchanged messages, creates a reliable, scalable and easy to understand model for handling state.
+
+### Summary of state at service level
+
+At this point, we've already covered variables, flow control statements (not really relevant for this post), functions, objects and threads. We focused on complexity more than scalability, but my thesis is that those ideas are aligned. To reduce the complexity derived from state at service level, we learned that the following properties are convenient:
+
+* Isolation
+* Avoiding state as much as possible
+* State should be managed by specialised software
+* Immutability
+* State and behaviour affinity
+
+Let's see how those ideas apply to system level. That corresponds to Distribution in the Learning Code list seen before.
+
+## State at system level
+
+State is data at a given instant in time. But what is the lifespan of that state? Request scoped? Session scoped? Could it be flushed out when a service crashes? Or should be persisted in a durable storage?
+
+* Request scoped: let's imagine an object holding a [Finite state machine](http://codurance.com/2016/05/10/finite-state-machines-with-akka/) during the lifespan of an HTTP request. This scope is really short so we could keep the state in memory. Most of the times we could simply retry on failures instead of using something more complicated like [Akka Persistence](http://doc.akka.io/docs/akka/current/scala/persistence.html).
+
+* Session scoped: HTTP is a stateless protocol. Remember that *time* is the key word in our definition of state. That means that HTTP doesn't have any built-in mechanism to keep track of state. The concept of session involves a logical group of HTTP requests with some internal state. Classic solution was keeping that state in memory on the service. This solution goes against horizontal scalability. The main point of having stateless services is being able to route requests to different services depending on health and current load of those services. That's not possible when every single service is coupled with the clients through that session state.
```

Comment:
I agree that is confusing @dhatanian I've rephrased it:

"* Race conditions: if two clients send updates onto the same logical record, there is no guarantee that those updates will be applied ordered into the different datastores. Isolation, from ACID properties, helps you with this concurrency problem, and implementing it at service level has some complexities."